dsl/src/main/scala/com.crobox.clickhouse/dsl/language/ClickhouseTokenizerModule.scala

```diff
@@ -214,70 +226,79 @@ trait ClickhouseTokenizerModule
         }
-        s"""${if (query.global) "GLOBAL " else ""}
+
+        s""" AS ${ctx.leftAlias}
+           | ${if (query.global) "GLOBAL " else ""}
            | ${tokenizeJoinType(query.joinType)}
-           | $other AS ${query.alias}
-           | ${tokenizeJoinKeys(from.get, query)}""".trim.stripMargin
+           | $right AS ${ctx.rightAlias}
+           | ${tokenizeJoinKeys(select, from.get, query)}""".trim.stripMargin
           .replaceAll("\n", "")
           .replaceAll("\r", "")
       case None => ""
     }
 
-  private def tokenizeJoinKeys(from: FromQuery, query: JoinQuery): String = {
-    val joinKeys = query.joinKeys.filterNot {
+  private def tokenizeJoinKeys(select: Option[SelectQuery], from: FromQuery, query: JoinQuery)(
+      implicit ctx: TokenizeContext
+  ): String = {
+
+    val using = query.using.filterNot {
       case EmptyColumn => true
       case _           => false
     }
     query.joinType match {
       case CrossJoin =>
-        assert(joinKeys.isEmpty, "When using CrossJoin, no joinKeys should be provided")
+        assert(using.isEmpty, "When using CrossJoin, no using columns should be provided")
+        assert(query.on.isEmpty, "When using CrossJoin, no on conditions should be provided")
         ""
       case _ =>
-        assert(joinKeys.nonEmpty, s"No joinKeys provided for joinType: ${query.joinType}")
-
-        val (aliasFrom, columns) = from match {
-          case tableJoin: TableFromQuery[_] => (tableJoin.table.name, tableJoin.table.columns)
-          case innerJoin: InnerFromQuery    => ("", Seq.empty)
-        }
-
-        val clause = "ON " + joinKeys.map(c => tokenizeJoinKey(c, columns, aliasFrom, query.alias)).mkString(" AND ")
-
-        // check for match conditions (ASOF)
-        query.joinType match {
-          case AsOfJoin | AsOfLeftJoin =>
-            assert(query.matchConditions.nonEmpty, s"No matchConditions provided for joinType: ${query.joinType}")
-            clause + " AND " + query.matchConditions
-              .map(tuple => tokenizeJoinKey(tuple._1, columns, aliasFrom, query.alias, tuple._2))
-              .mkString(" AND ")
-          case _ =>
-            // fallthrough
-            clause
-        }
+        assert(using.nonEmpty || query.on.nonEmpty, s"No USING or ON provided for joinType: ${query.joinType}")
+        assert(!(using.nonEmpty && query.on.nonEmpty), s"Both USING and ON provided for joinType: ${query.joinType}")
+
+        if (using.nonEmpty) {
+          // TOKENIZE USING
+          if (using.size == 1) s"USING ${using.head.name}"
+          else s"USING (${using.map(_.name).mkString(",")})"
+        } else if (query.on.nonEmpty) {
+          // TOKENIZE ON. If the fromClause is a TABLE, we need to check on aliases!
+          "ON " + query.on
+            .map(cond => {
+              val left = verifyOnCondition(select, from, cond._1)
+              s"${ctx.leftAlias}.$left ${cond._2} ${ctx.rightAlias}.${cond._3.name}"
```

Comment:
cond should be a case class instead of a Tuple3; now it's not really clear what's what.

Reply:
Added JoinCondition case class